### DISCUSSION TITLE
SoftI2C primitive operations

### DISCUSSION THREAD

**Discussion by riteman** [micropython/micropython#9520] [General]
Hello
I am working on a project where I need to access (read/write) data to an Artmel 24C32n eeprom. I have tried different libraries (mcauser and a modified version of peterhinch') but none of them work for me so I decided to use I2C primitives and develop a simple driver based on the datasheet from Atmel.
I read that only SoftI2C supports primitive operations which is the reason I import and use it.
I use Thonny 4.0.1 and micropython 1.19.1 running on a Raspberry Pi Pico W.

Here a snippet where I only read/write one byte:
```py
import machine, time, ds1307
from machine import SoftI2C

i2c_addr = 0x57
i2c= SoftI2C(machine.Pin(1), machine.Pin(0), freq=100000, timeout = 50000) #max clock @ 3.3V
devices = i2c.scan()
print(len(devices))
for device in devices:
   print("Decimal address: ",device," | Hexa address: ",hex(device))

def EEWriteByte(address, bytedata):
    i2c.start()
    i2c.write(i2c_addr << 1) # add a write bit at the end
    NoOfACK = i2c.write(address >> 8)   #upper byte
    NoOfACK = i2c.write(address & 0xff) #lower byte
    NoOfACK = i2c.write(bytedata)
    i2c.stop()
    time.sleep_ms(20)    
    
def EEReadByte(address,buf):
    i2c.start()
    i2c.write(i2c_addr << 1)                       # add a write bit at the end
    NoOfACK = i2c.write(address >> 8)    #upper byte
    NoOfACK = i2c.write(address & 255)  #lower byte
    i2c.stop()
    i2c.start()
    i2c.write(i2c_addr << 1 + 1)               # add a read bit at the end
    i2c.readinto(buf, nack=True)
    i2c.stop()
    time.sleep_ms(20)
    return buf
    
buf = bytearray(1)
buf[0] = 0x55
print(buf)
EEWriteByte(0,buf)
print(EEReadByte(0,buf))  
```
When I try to run this snippet, I receive an error

Traceback (most recent call last):
  File "<stdin>", line 36, in <module>
  File "<stdin>", line 13, in EEWriteByte
TypeError: object with buffer protocol required

The GO  LE keys on my keyboard almost are worn out and still I have not been able to find the reason.

Any hint which can make this work ?

Regards

**Comment by robert-hh** [comment]
Lines like these `i2c.write(i2c_addr << 1)` are the reason, because you supply an integer as argument, but the call requires a buffer type argument, like a bytearray or bytes object. The same for the two following calls.  You could use a bytearray of size 3, fill it with the data and then have  a single write. Or use uctypes.pack, or ....

**Reply by riteman** [reply]
Thanks a lot 

- "size 3" because byte 1= high address, byte 2 = low address and byte 3 = data?
But how do I force the I2C-address to become a byte?
Data I use in the call is a bytearray but that seems to be ignored in the subroutine?

uctype would that be something like "i2c_addr": uctypes.UINT8 | 0x57 which I guess ensures that the chip address is a byte?
I did not find any references to uctype.pack?

**Reply by robert-hh** [reply]
If data is a single byte as well, you will need a bytearray of size 4, for I2C-address, 2 byte FRAM address and data. Since data is a bytearray already, you it stays with 3 bytes of I2C-Address and FRAM address, and anther write call for the data.
But I have doubts on whether that would change the result compared to using i2c.writeto_mem(), because you code the same thing in Python instead of C.

**Reply by riteman** [reply]
using a bytearray I still get the same error:

```py
import machine, time, ds1307
from machine import SoftI2C

i2c_addr = 0x57
i2c= SoftI2C(machine.Pin(1), machine.Pin(0), freq=100000, timeout = 50000) #max clock @ 3.3V
devices = i2c.scan()
print(len(devices))
for device in devices:
   print("Decimal address: ",device," | Hexa address: ",hex(device))

def EEWriteByte(buf):
    i2c.start()
    i2c.write(buf[0] << 1) # add a write bit at the end
    NoOfACK = i2c.write(buf[1]) #upper byte
    NoOfACK = i2c.write(buf[2]) #lower byte
    NoOfACK = i2c.write(buf[4])
    i2c.stop()
    time.sleep_ms(20)    
    
def EEReadByte(buf):
    i2c.start()
    i2c.write(buf[0]) # add a write bit at the end
    NoOfACK = i2c.write(buf[1]) #upper byte
    NoOfACK = i2c.write(buf[2]) #lower byte
    i2c.stop()
    i2c.start()
    i2c.write(buf[0] << 1 + 1) # add a read bit at the end
    i2c.readinto(buf[3], nack=True)
    i2c.stop()
    time.sleep_ms(20)
    return buf
    
buf = bytearray(5)
buf[0] = 0x57  # i2c_address
buf[1] = 0x00  # high address of memory
buf[2] = 0x00  # low address of memory
buf[3] = 0xaa  # byte to write
print(buf)
EEWriteByte(buf)
print(EEReadByte(buf[3]))    
```

**Reply by robert-hh** [reply]
just call 
i2c.write(buf)
to send all 3 bytes at once. buf[0] etc. is again an integer.

**Reply by riteman** [reply]
How does micropython know that "buf" in the subroutine is a bytearray? I am new to Python - in other languages a subroutine would look like this 
NameOfRoutine(data : Integer, buf: Bytearray)

**Reply by robert-hh** [reply]
Python carries along information of the type with the object. For a bytearray object, it is the size and the type (and maybe more). But if buf is a bytearray, then buf[0] is not, it's an integer. However, buf[0:1] is a bytearray again.

About the error: I still see calls to write with integers. Like in the second-to-last line, you still call EEWriteByte(buf), which calls write. Maybe this variant of the code works. I have no appropriate hardware here to test. But at least on the bus this shows the right sequence.
```
import machine, time, ds1307
from machine import SoftI2C

i2c_addr = 0x57
i2c= SoftI2C(sda=machine.Pin("SDA"), scl=machine.Pin("SCL"), freq=100000, timeout = 50000) #max clock @ 3.3V
devices = i2c.scan()
print(len(devices))
for device in devices:
   print("Decimal address: ",device," | Hexa address: ",hex(device))

def EEWriteByte(address, bytedata):
    buf3 = bytearray(3)
    buf3[0] = i2c_addr << 1 # add a write bit at the end
    buf3[1] = address >> 8  #upper byte
    buf3[2] = address & 0xff  #lower byte
    i2c.start()
    NoOfACK = i2c.write(buf3 + bytedata)
    i2c.stop()
    time.sleep_ms(20)

def EEReadByte(address, data):
    buf3 = bytearray(3)
    buf3[0] = i2c_addr << 1 # add a write bit at the end
    buf3[1] = address >> 8  #upper byte
    buf3[2] = address & 0xff  #lower byte
    i2c.start()
    i2c.write(buf3)
    i2c.stop()
    i2c.start()
    buf3[0] += 1   # add a read bit at the end
    i2c.write(buf3[0:1]) # just write the first byte
    i2c.readinto(data)
    i2c.stop()
    time.sleep_ms(20)
    return data

buf = bytearray(1)
buf[0] = 0x55
print(buf)
EEWriteByte(0,buf)
print(EEReadByte(0, buf))
```

**Reply by riteman** [reply]
I modified the snippet as shown below

```py
import machine, time, ds1307
from machine import SoftI2C

i2c_addr = 0x57
i2c= SoftI2C(machine.Pin(1), machine.Pin(0), freq=100000, timeout = 50000) #max clock @ 3.3V
devices = i2c.scan()
print(len(devices))
for device in devices:
   print("Decimal address: ",device," | Hex address: ",hex(device))

def EEWriteByte(buf):
    i2c.writeto_mem(i2c_addr, 4, buf)  #write buf into eeprom from addr 4
    time.sleep_ms(20)    
    
def EEReadByte():
    i2c.readfrom_mem_into(i2c_addr, 4, buf)  #read len(buf) bytes into buf
    return buf
    
buf = bytearray(4)
print(EEReadByte())
buf[0] = 0xaa  
buf[1] = 0x55  #uppercase U
buf[2] = 0xaa 
buf[3] = 0x55  # uppercase U
print('Data being written:', buf)
EEWriteByte(buf)
print('Data from EEPROM following write:',EEReadByte())
```
The results are 
bytearray(b'xff0')
Data being written: bytearray(b'\xaaU\xaaU')
Data from EEPROM following write: bytearray(b'xff0')

thus the write did not succeed and the number of bytes read is not 4?? 
(I do have pull ups on SDA and SCL)

**Reply by riteman** [reply]
I really appreciate the time taken to help me and your comments on the Python way of transferring info between main and subroutine help a lot - I will try your snippet!

**Reply by riteman** [reply]
..unfortunately th eresults are 

0
bytearray(b'U')
bytearray(b'\xff')

**Reply by riteman** [reply]
BTW: How did you check the bus? Storagescope ?

**Reply by jimmo** [reply]
> Data from EEPROM following write: bytearray(b'xff0')

This is saying that the bytearray contained four bytes -- `['0x78', '0x66', '0x66', '0x30']`.  (Note that there are no backslashes in the bytearray printout, it literally contains the bytes "x", "f", "f", "o").

I took a quick look at the datasheet, the register addresses for this device are 16-bit.

So you should be setting addrsize=16 on `writeto_mem` and `readfrom_mem_into`.

**Reply by robert-hh** [reply]
> BTW: How did you check the bus? Storagescope ?

Simple 10$ USB interface logic analyzer.

The first line with 0 in the printout tells, that there is no connection to the I2C device. So the wiring is wrong. Unless that is fixed, you will not succeed.
- Did you connect all wires to the right pins? GND, Vcc, SDA and SCL?

b.t.w.: Deep in my drawer I had a DS3231 breakout with an EEPROM chip. Using that, writing & reading  worked.
Using i2c.writeto_mem() and readfrom_mem_into() failed. That's strange.

**Reply by robert-hh** [reply]
i2c.writeto_mem() and readfrom_mem_into() work now. It required a sleep after write. So that's the script now:
```
import machine, time, ds1307
from machine import SoftI2C

i2c_addr = 0x57
i2c= SoftI2C(sda=machine.Pin("SDA"), scl=machine.Pin("SCL"), freq=100000, timeout = 50000) #max clock @ 3.3V
devices = i2c.scan()
print(len(devices))
for device in devices:
   print("Decimal address: ",device," | Hexa address: ",hex(device))

buf = bytearray(1)
buf[0] = 0x63
print(buf)
i2c.writeto_mem(i2c_addr, 0x0003, buf, addrsize=16)
time.sleep_ms(20)
i2c.readfrom_mem_into(i2c_addr, 0x0003, buf, addrsize=16)
print(buf)
```

**Reply by riteman** [reply]
What does the addrsize = 16 mean. I do not use that parameter and it works fine

**Reply by robert-hh** [reply]
It tells, that the memory address is 16 bits long. You do not use it because you do not use i2c.writeto_mem() or i2c.readfrom_mem_into(). You have written your own versions of these methods.

**Reply by riteman** [reply]
Thanks!
I will try your version as well,,..

**Reply by riteman** [reply]
after some slight modifications it worked fine:

```py
import machine, time
from machine import SoftI2C

i2c_addr = 0x57
i2c= SoftI2C(machine.Pin(1), machine.Pin(0), freq=100000, timeout = 50000) #max clock @ 3.3V
devices = i2c.scan()
print(len(devices))
for device in devices:
   print("Decimal address: ",device," | Hex address: ",hex(device))

buf = bytearray(1)
buf[0] = 0x63
print(buf)
i2c.writeto_mem(i2c_addr, 0x0003, buf, addrsize=16)
time.sleep_ms(20)
i2c.readfrom_mem_into(i2c_addr, 0x0003, buf, addrsize=16)
print(buf)

``

**Comment by riteman** [comment]
Hi

I had no problem with the wiring and the SDA/SCL pinning was correct 
from the beginning or I would not have been able to scan the bus.
The info about data from a bytearray being seen as an integer solved the 
problem - Using the type command I spent some time checking the 
difference between Buf[0] and Buf[0:1] ;-)
The chip is an 8 bit eeprom and the address being read/written is split 
into 2 bytes and sent MSB followed by LSB

Here the current version:

```py

# MIT License
# part of the code Copyright (c) 2018 Mike Causer
# The wait routine is Copyright (c) 2019  Peter Hinch
# Slightly modified

import machine, time
from machine import SoftI2C

i2c= SoftI2C(machine.Pin(1), machine.Pin(0), freq=100000, timeout = 
50000) #max clock @ 3.3V
  100 kHz
i2c_addr   = 0x57
pages       = 128
bpp          = 32               # bytes per page
WriteByte  = 2*i2c_addr + 0   # this is the device address including the 
write bit  as LSB (=0)
ReadByte   = 2*i2c_addr + 1   # and this is the device address with a 
Read bit (=1)
WriteBytes = bytearray(3)
Buf1           = bytearray(1)

devices    = i2c.scan()

print('Number of units found: ',len(devices))
for device in devices:
    print("Decimal address: ",device," | Hex address: ",hex(device))

def Wait4ready():  # After a write, wait for device to become ready
         Buf1[0] = 0
         while True:
             try:
                 if i2c.writeto(i2c_addr, Buf1):  # Poll ACK
                     break
             except OSError:
                 pass
             finally:
                 time.sleep_ms(1)

def read(addr, nbytes):
     RollOver = ((addr + nbytes) > 4094)  #if we pass the 4 k boundary, 
the address-counter will roll over and start at address 0
     """Read one or more bytes from the EEPROM starting from a specific 
address"""
     WriteBytes[0] = WriteByte
     WriteBytes[1] = addr >> 8  # first MSB is being sent out
     WriteBytes[2] = addr & 0xff  # the LSb is sent
     i2c.start() # start a dummy write to set the address counter in the 
eeprom
     i2c.write(WriteBytes) # make a dummy write request (no STOP)
     Res = i2c.readfrom(i2c_addr, nbytes) # return data
     return Res

def write(addr, buf):
     """Write one or more bytes to the EEPROM respecting the block 
boundaries """
     RollOver = addr + len(buf) > 4095
     offset = addr % bpp
     partial = 0
     # partial page write
     if offset > 0:
         partial = bpp - offset
         WriteBytes[2] = addr & 0xff          #isolate LSB of addr
         WriteBytes[1] = (addr >> 8) & 0xff   #and MSB
         WriteBytes[0] = WriteByte       #controlword for addressing 
EEPROM for write operations
         i2c.start()
         i2c.write(WriteBytes + buf[0:partial])
         i2c.stop()
         #Now wait for the eeprom to finish the write cyclus
         Wait4ready()
         #time.sleep_ms(20)
         addr += partial
     # full page write
     for i in range(partial, len(buf), bpp):
         WriteBytes[2] = (addr + i-partial) & 0xff            #isolate 
LSB of addr
         WriteBytes[1] = ((addr + i - partial) >> 8) & 0xff   #and MSB
         WriteBytes[0] = WriteByte                       #controlword for 
addressing EEPROM for write operations
         i2c.start()
         i2c.write(WriteBytes + buf[i:i+bpp])
         i2c.stop()
         Wait4ready()
         #time.sleep_ms(20)
     return RollOver

EEprom = read(0,8)
print('8 bytes read before write takes place: ', EEprom)
Rollover = write(0,'Sed ut perspiciatis unde omnis iste natus error sit 
voluptatem accusantium doloremque laudantium  ')
print('Rollover during write operation? ', Rollover)
EEprom = read(0,128)
print('128 byte read: ', EEprom)

```
which results in the following output:

Number of units found:  3
Decimal address:  60  | Hex address:  0x3c
Decimal address:  87  | Hex address:  0x57
Decimal address:  104  | Hex address:  0x68
8 bytes read before write takes place:  b'Sed ut p'
Rollover during write operation?  False
128 byte read:  b'Sed ut perspiciatis unde omnis iste natus error sit 
voluptatem accusantium doloremque laudantium  
cdefghijklmnopqrstuvwxyz012345'

Thanks for all the good answers and amount of work invested. It's very 
much appreciated!

Regards

**Comment by hardikofficial963** [comment]
**Issue:**
 I2C Scan on ESP32-S3 Returns No Devices

**Description:**
I am trying to scan for I2C devices on the ESP32-S3 using MicroPython but getting no devices detected, even though the hardware works with the Arduino Wire Scan example.

**MicroPython Version:**
v1.24.0 (2024-10-25)

**Hardware Details:**
- Microcontroller: ESP32-S3
- I2C Device: PCA9685
- Pins Used:
     - SCL: GPIO 19
     - SDA: GPIO 20

**Code:**

        from machine import Pin, SoftI2C
        import time
        
        def scan_i2c_addresses():
            # Initialize I2C
            i2c = SoftI2C(sda=Pin(19), scl=Pin(20), freq=400000)
            
            print("Scanning I2C addresses...")
            devices = i2c.scan()
            
            if len(devices) == 0:
                print("No I2C devices found!")
            else:
                print(f"Found {len(devices)} I2C devices:")
                for device in devices:
                    print(f"Device address: {hex(device)}")
        
        if __name__ == '__main__':
            scan_i2c_addresses()

    
**Observations:**

- Expected Behavior: Detect the PCA9685 at address 0x40.
- Actual Behavior: "No I2C devices found!" message is printed.

**Troubleshooting Performed**
1. Tested the same hardware and wiring configuration using the Arduino Wire Scan example. The device at 0x40 is detected successfully.
2. Verified the wiring connections:
    SCL → GPIO 19
    SDA → GPIO 20
3. Proper pull-up resistors in place.

**Request:**

- Is there a known issue with SoftI2C on ESP32-S3 in MicroPython?
- Are there any additional configurations or debugging steps I should try?

Any insights would be appreciated.

![Screenshot 2024-11-21 174820](https://github.com/user-attachments/assets/fd76d7d8-f8ae-4580-8220-322a52a972a3)

**Reply by peterhinch** [reply]
Please see [my response](https://github.com/micropython/micropython/issues/16283#issuecomment-2493393151) to your issue in the MP bug tracker.
